docs/rfcs/xxxx-backwards-compatibility.md

```diff
@@ -0,0 +1,179 @@
+
+If a breaking change is introduced in a minor change accidentally, and that
+breaking change is not covered by one of the [exceptions][] defined below, it
+is a bug. In these cases, a patch release should be introduced to undo the
+breaking change. 
+
+[exceptions]: #exceptions-to-backwards-compatibility
+
+### Exceptions to backwards compatibility 
+
+It's impossible to guarantee that full backwards compatibility is achieved.
+There are some exceptions which may cause a breaking change without a new major
+version:
+
+- Security: a breaking change may be made if a security fix requires making a
+  breaking change. 
+
+- Non-versioned network APIs: internal network APIs, such as the internal API
+  used to drive the Flow web UI, are not subject to backwards compatibility
+  guarantees.
+
+- Undocumented behavior: relying on undocumented behavior may break between
+  minor releases. 
+
+- Upstream dependencies: part of the public API of Grafana Agent may directly
+  expose the public API of an upstream dependency. In these cases, if an
+  upstream dependency introduces a breaking change, we may be required to make
+  a breaking change to our public API as well.   
+
+- Other telemetry data: metrics, logs, and traces may change between releases.
+  Only telemetry data which is used in official dashboards is protected under
+  backwards compatibility.
+
+### Avoiding breaking changes 
+
+Backwards compatibility guarantees introduce new challenges while writing code,
+as it means finding creative ways to iterate without making breaking changes
+immediately. Here are some example strategies one could use to avoid making a
+breaking change:
+
+- Keep things internal: the easiest route is to just not expose things
+  publicly. For example, put Go code in `internal/` packages, or use unexported
+  methods or types.
+
+- Rely on deprecations: if a publicly exposed thing must be changed, consider
+  deprecating it.
+
+  - Instead of removing something, consider making it a no-operation and adding
+    a deprecation warning. 
+
+  - Instead of replacing something (Go methods, Flow components, Flow attribute
+    names), consider introducing the new approach and deprecating the old
+    approach. 
+
+A list of deprecations will accumulate over the course of a major version's
+lifecycle. This list must be tracked so that deprecated functionality can be
+removed with the next major version. 
+
+### Avoiding major release burnout 
+
+As a new major release implies a user must put extra effort into upgrading, it
+is possible to burn out users by releasing breaking changes too frequently. 
+
+We will attempt to limit new major versions no more than once every 12 calendar
+months. This means that if Grafana Agent 1.0 was hypothetically released on
+August 4th, Grafana Agent 2.0 should not be released until at least August 4th
+of the following year. This is best-effort; if a new major release is required
+earlier, then one should be published. 
```

Comment:
I reckon that we'd have a batch of breaking changes for each major release and they would need to be parked on some branch and would be available to the public anyway. Maybe we should define what 'released' and 'published' mean?

Reply:
I've tried to clarify this 👍